### ISSUE TITLE
SymbolFinder.FindReferencesAsync TaskCanceledException wrapped as AggregateException 

### ISSUE THREAD

**Issue by pieroci** [dotnet/roslyn#12872]
**Version Used**: 
1.3.2

Visual Studio 2015 Update 3

**Steps to Reproduce**:

Hi all,
I have a framework solution referenced by an application solution (more than 80 projects and 3kk lines of code) .
I want all the references in **Application Solution**  (_solution_) of a particoular Method class( "_methodSymbol_" ) of the **Framework Solution** : to achieve this goal I pass _all the document_ (_project.Documents.ToImmutableHashSet<Document>()_) present in every single project of the Application Solution () .

Writing this code I obtain an Aggregate exception( inner exception : task was cancelled ) :

`var tsk= SymbolFinder.FindReferencesAsync(_methodSymbol_, _solution_, project.Documents.ToImmutableHashSet<Document>(), new System.Threading.CancellationToken());
var allReferences = tsk.Result
`
Call Stack:

>  at System.Threading.Tasks.Task.ThrowIfExceptional(Boolean includeTaskCanceledExceptions)
>    at System.Threading.Tasks.Task`1.GetResultCore(Boolean waitCompletionNotification)
>    at System.Threading.Tasks.Task`1.get_Result()
>    at RoslynUtility.WorkspaceSolution.GetAllRefByMethodSymbol(Solution solution, List`1 result, IMethodSymbol methodSymbol, Project project) in C:\Users\xxxx\Documents\Visual Studio 2015\Projects\RoslynUtility\RoslynUtility\Program.cs:line 202

I try to return successfully all the references for some projects of the solution, except the one that throws this exception. I try, thus, to Parse every single Document of this project:

```
...
foreach(var documentId in project.DocumentIds)
{

Document document = solution.GetDocument(documentId);
var singleClassOfProject = project.Documents.ToImmutableHashSet<Document>().Where(x => x.Id == documentId).ToImmutableHashSet();

var tsk = SymbolFinder.FindReferencesAsync(methodSymbol, solution, singleClassOfProject, new System.Threading.CancellationToken());
}

```

Call Stack:

>    at System.Threading.Tasks.Task.ThrowIfExceptional(Boolean includeTaskCanceledExceptions)
>    at System.Threading.Tasks.Task`1.GetResultCore(Boolean waitCompletionNotification)
>    at System.Threading.Tasks.Task`1.get_Result()
>    at RoslynUtility.WorkspaceSolution.TryParseSingleDocument(Solution solution, List`1 result, IMethodSymbol methodSymbol, Project project) in C:\Users\xxxxxx\Documents\Visual Studio 2015\Projects\RoslynUtility\RoslynUtility\Program.cs:line 170

With this use of Symbol Finder  I can't parse all the documents (same problem : Aggregate Exception ; Task was cancelled).

**Expected Behavior**:

parse all project

**Actual Behavior**:

Aggregate Exception; Inner Exception: "task was cancelled"

**Comment by Pilchie** [comment]
@pieroci Can you post the call stack of the TaskCancelledException?

@CyrusNajmabadi - any ideas here?

**Comment by pieroci** [comment]
@Pilchie  
I've integrated the post with the call stack, but it's not very explanatory

**Comment by CyrusNajmabadi** [comment]
Do you have a more complete sample that i can use?

**Comment by CyrusNajmabadi** [comment]
> to achieve this goal I pass all the document (project.Documents.ToImmutableHashSet()) present in every single project of the Application Solution () .

Note: the default behavior is to search everything int he solution.  You don't need to pass along any documents to get this behavior.  You only pass documents if you want to *limit* the set that is searched.

**Comment by pieroci** [comment]
> Do you have a more complete sample that i can use?

yes, I can send you this little Roslyn project, but not the Solution and Framework Solution for privacy company policies. This Roslyn Project worked in a test-solution-Framework and in a Test-Application-Solution.

> Note: the default behavior is to search everything int he solution. You don't need to pass along any documents to get this behavior. You only pass documents if you want to limit the set that is searched.

If I try to retrieve all references of my entire solution I obtain the same Exception:

```
   at System.Threading.Tasks.Task.ThrowIfExceptional(Boolean includeTaskCanceledExceptions)
   at System.Threading.Tasks.Task`1.GetResultCore(Boolean waitCompletionNotification)
   at System.Threading.Tasks.Task`1.get_Result()
   at RoslynUtility.WorkspaceSolution.GetAllReferenceByMethodName(String methodNameToSearch, List`1 documents, Solution solution) in C:\Users\xxxxxxx\Documents\Visual Studio 2015\Projects\RoslynUtility\RoslynUtility\Program.cs:line 140
```

**Comment by CyrusNajmabadi** [comment]
You mentioned you're getting an aggregate exception.  Can you catch that exception and print out the stacks of all its [InnerExceptions](https://msdn.microsoft.com/en-us/library/system.aggregateexception.innerexceptions(v=vs.110).aspx)?  Thanks!

**Comment by pieroci** [comment]
yes, thank you .


`e.InnerExceptions
Count = 1
    [0]: {"A task was canceled."}
`

The call Stack is Null.
`		(e.InnerExceptions[0]).StackTrace	null	
`

**Comment by CyrusNajmabadi** [comment]
That's unfortunate.  Can you collect a dump when this happens, upload it somewhere, and then send us a link?  I'd definitely like to look at what's going on.

**Comment by pieroci** [comment]
I shared it with [OneDrive](https://avanade-my.sharepoint.com/personal/piero_ciula_avanade_com/Documents/Dump/RoslynUtility.vshost.dmp).
tell me if you can access or not .I send you yet an email-invitation. 
Thank you so mutch!

**Comment by CyrusNajmabadi** [comment]
Downloading now!

**Comment by CyrusNajmabadi** [comment]
Hrmm... dump is not illuminating.  Can we have your mini test program you've written?  I'd like to run it against something the size of Roslyn to see if i can observe hte same behavior.  Thanks!

**Comment by pieroci** [comment]
Ok I shared you , and send you an email invitation.
You can find in Program directory: TestFrameworkSolution , TestApplicationSolution and RoslynUtility solution, but I hope Help you. :)
RoslynUtility worked with TestFrameworkSolution and TestApplicationSolution .
Instructions: Copy "Program" directory in C:\ and Launch RoslynUtility solution.

tell me if you can access or not . :)
Thank you so mutch!

**Comment by CyrusNajmabadi** [comment]
Hate to be a bother.  However, OneDrive is not cooperating.  Can you zip these into a single file to make it easier to download?

**Comment by CyrusNajmabadi** [comment]
Ok, i've been running this a while on my box with no exceptions thrown so far.  When you debug locally do you see any first chance exceptions thrown at all?

**Comment by CyrusNajmabadi** [comment]
Note: if you do get an interesting first-chance exception of any sort, please take a dump of that.  Thanks!

**Comment by pieroci** [comment]
With the Test Application and the Test Framework Solutions I don't get any sort of exception. The exception comes out only when I use RoslynUtility on a particular project of my End Costumer Solution Project.

I think that the problem in the Solution Project is caused by this particular project references.

**Comment by CyrusNajmabadi** [comment]
> I think that the problem in the Solution Project is caused by this particular project references.

are you able to run this under the debugger with first-chance exceptions enabled?  it would be valuable to know if you're hitting anything.

**Comment by CyrusNajmabadi** [comment]
Any updates on this?  Have you been able to see any sort of first chance exception?  Thanks!

**Comment by lukehunter** [comment]
I am having a similar problem (Document.GetSemanticModelAsync always fails with AggregateException containing a single TaskCanceledException with a null stacktrace). I have some projects for which this call fails on every document. I am also using roslyn 1.3.2 and VS2015 Update 3.

If I change the code to call Project.GetCompilationAsync and turn on first chance exceptions, VS breaks on the following exceptions for the same projects that fail with Document.GetSemanticModelAsync:
System.InvalidCastException in Microsoft.Build.dll
Additional information: Unable to cast object of type 'Microsoft.Build.Tasks.Windows.MarkupCompilePass1' to type 'Microsoft.Build.Framework.ITask'.

This is followed by two throws of Microsoft.Build.Exceptions.InvalidProjectFileException in Microsoft.Build.dll and in mscorlib.dll with the message:
The "MarkupCompilePass1" task has been declared or used incorrectly, or failed during construction. Check the spelling of the task name and the assembly name.

I tried re-creating one of the offending projects in VS2015 (most of these projects are old and have been converted through multiple VS versions) but it didn't help. Unfortunately I can't share my code since it is proprietary but I have attached an anonymized version of one of the failing project files and an anonymized version of my messy roslyn experimentation code.

I also cross checked all the MetadataReferences from the bad project files, but the only references that were in all the failing projects were also present in at least one of the succeeding projects. I also couldn't find a correlation between WPF dependent projects or between class library vs windows application projects.

Hope this helps, let me know if there is any more information I can provide. 

[anonymized.csproj.txt](https://github.com/dotnet/roslyn/files/444312/anonymized.csproj.txt)

[anonymized_roslyn.cs.txt](https://github.com/dotnet/roslyn/files/444347/anonymized_roslyn.cs.txt)

**Comment by Pilchie** [comment]
Hi @lukehunter for the msbuild exceptions you're seeing, you'll probably need to add binding redirects for the msbuild assemblies to your app.config file.  See https://github.com/dotnet/roslyn/issues/2779 for more details.

**Comment by pieroci** [comment]
@CyrusNajmabadi 

> 
> CyrusNajmabadi commented 18 days ago
> I think that the problem in the Solution Project is caused by this particular project references.
> are you able to run this under the debugger with first-chance exceptions enabled? it would be valuable to know if you're hitting anything.
> @CyrusNajmabadi
>  
> CyrusNajmabadi commented 6 days ago
> Any updates on this? Have you been able to see any sort of first chance exception? Thanks!


Hi, unfortunately the exception that I showed you , is chatched with first chance exception enabled . 
I downloaded the roslyn sourcecode for an eventually test in debug. 

I will try to send the new update , if I can set it up properly with Application that throws this exception.
Thanks again.

**Comment by CyrusNajmabadi** [comment]
@pieroci Any movement on this? Thanks!

**Comment by KirillOsenkov** [comment]
Try creating a solution with a duplicate project reference, see if you can hit it that way.

**Comment by barispeker** [comment]
Hello I had the same problem. Removing duplicate project references solved the problem. Still unfortunately finding this solution cost me 2 days.

**Comment by pieroci** [comment]
Hi all, beg your pardon for answering so late.
I got the Roslyn source code and debugged the exceptions. 
The problem was not due to the duplicate references and wasn't due to SymbolFinder.FindReferencesAsync method, but was a build issue. Calling the OpenSolutionAsync the solution wasn't compiled correctly and there were no exceptions showed in release mode.
I resolved setting the version 14 of assembly Microsoft.Build.Framework in **msbuild.exe.config** ( C:\Program Files (x86)\MSBuild\14.0\bin).

```

<dependentAssembly>
    <assemblyIdentity name="Microsoft.Build.Framework" publicKeyToken="b03f5f7f11d50a3a" culture="neutral"/>
    <bindingRedirect oldVersion="0.0.0.0-99.9.9.9" newVersion="14.0.0.0"/>
</dependentAssembly>

```
